// @blitzjs/auth/dist/index-browser.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./chunks/index.cjs');
require('b64-lite');
require('bad-behavior');
require('react');
require('blitz');
require('debug');



exports.AuthClientPlugin = index.AuthClientPlugin;
exports.COOKIE_ANONYMOUS_SESSION_TOKEN = index.COOKIE_ANONYMOUS_SESSION_TOKEN;
exports.COOKIE_CSRF_TOKEN = index.COOKIE_CSRF_TOKEN;
exports.COOKIE_PUBLIC_DATA_TOKEN = index.COOKIE_PUBLIC_DATA_TOKEN;
exports.COOKIE_REFRESH_TOKEN = index.COOKIE_REFRESH_TOKEN;
exports.COOKIE_SESSION_TOKEN = index.COOKIE_SESSION_TOKEN;
exports.HANDLE_SEPARATOR = index.HANDLE_SEPARATOR;
exports.HEADER_CSRF = index.HEADER_CSRF;
exports.HEADER_CSRF_ERROR = index.HEADER_CSRF_ERROR;
exports.HEADER_PUBLIC_DATA_TOKEN = index.HEADER_PUBLIC_DATA_TOKEN;
exports.HEADER_SESSION_CREATED = index.HEADER_SESSION_CREATED;
exports.LOCALSTORAGE_CSRF_TOKEN = index.LOCALSTORAGE_CSRF_TOKEN;
exports.LOCALSTORAGE_PREFIX = index.LOCALSTORAGE_PREFIX;
exports.LOCALSTORAGE_PUBLIC_DATA_TOKEN = index.LOCALSTORAGE_PUBLIC_DATA_TOKEN;
exports.SESSION_TOKEN_VERSION_0 = index.SESSION_TOKEN_VERSION_0;
exports.SESSION_TYPE_ANONYMOUS_JWT = index.SESSION_TYPE_ANONYMOUS_JWT;
exports.SESSION_TYPE_OPAQUE_TOKEN_SIMPLE = index.SESSION_TYPE_OPAQUE_TOKEN_SIMPLE;
exports.TOKEN_SEPARATOR = index.TOKEN_SEPARATOR;
exports.backupAntiCSRFTokenToLocalStorage = index.backupAntiCSRFTokenToLocalStorage;
exports.getAntiCSRFToken = index.getAntiCSRFToken;
exports.getAuthValues = index.getAuthValues;
exports.getPublicDataStore = index.getPublicDataStore;
exports.parsePublicDataToken = index.parsePublicDataToken;
exports.useAuthenticatedSession = index.useAuthenticatedSession;
exports.useAuthorize = index.useAuthorize;
exports.useAuthorizeIf = index.useAuthorizeIf;
exports.useRedirectAuthenticated = index.useRedirectAuthenticated;
exports.useSession = index.useSession;
