// @blitzjs/rpc/dist/chunks/loader-utils.cjs
'use strict';

const blitz = require('blitz');
const path = require('path');

function assertPosixPath(path$1) {
  const errMsg = `Wrongly formatted path: ${path$1}`;
  blitz.assert(!path$1.includes(path.win32.sep), errMsg);
}
function toPosixPath(path$1) {
  if (process.platform !== "win32") {
    blitz.assert(path.sep === path.posix.sep, "TODO");
    assertPosixPath(path$1);
    return path$1;
  } else {
    blitz.assert(path.sep === path.win32.sep, "TODO");
    const pathPosix = path$1.split(path.win32.sep).join(path.posix.sep);
    assertPosixPath(pathPosix);
    return pathPosix;
  }
}
const topLevelFoldersThatMayContainResolvers = ["src", "app", "integrations"];
function buildPageExtensionRegex(pageExtensions) {
  return new RegExp(`(?<!\\.test|\\.spec)\\.(?:${pageExtensions.join("|")})$`);
}
const fileExtensionRegex = /\.([a-z]+)$/;
function convertPageFilePathToRoutePath({
  absoluteFilePath,
  resolverBasePath,
  appRoot,
  extraRpcBasePaths = []
}) {
  let path$1 = path.normalize(absoluteFilePath);
  if (typeof resolverBasePath === "function") {
    path$1 = resolverBasePath(path$1);
  } else if (resolverBasePath === "root") {
    path$1 = path$1.replace(path.normalize(appRoot), "");
    for (const extraPath of extraRpcBasePaths) {
      path$1 = path$1.replace(path.join(path.normalize(appRoot), extraPath.replace("/", path.sep)), "");
    }
  } else {
    path$1 = path$1.replace(/^.*?[\\/]queries[\\/]/, "/").replace(/^.*?[\\/]mutations[\\/]/, "/");
  }
  return path$1.replace(/\\/g, "/").replace(fileExtensionRegex, "");
}
function convertFilePathToResolverName(filePathFromAppRoot) {
  return filePathFromAppRoot.replace(/^.*[\\/](queries|mutations)[\\/]/, "").replace(fileExtensionRegex, "");
}
function convertFilePathToResolverType(filePathFromAppRoot) {
  return filePathFromAppRoot.match(/[\\/]queries[\\/]/) ? "query" : "mutation";
}
function getIsRpcFile(filePathFromAppRoot) {
  return /[\\/]queries[\\/]/.test(filePathFromAppRoot) || /[\\/]mutations[\\/]/.test(filePathFromAppRoot);
}

exports.assertPosixPath = assertPosixPath;
exports.buildPageExtensionRegex = buildPageExtensionRegex;
exports.convertFilePathToResolverName = convertFilePathToResolverName;
exports.convertFilePathToResolverType = convertFilePathToResolverType;
exports.convertPageFilePathToRoutePath = convertPageFilePathToRoutePath;
exports.getIsRpcFile = getIsRpcFile;
exports.toPosixPath = toPosixPath;
exports.topLevelFoldersThatMayContainResolvers = topLevelFoldersThatMayContainResolvers;
