// @blitzjs/next/dist/index-browser.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('next/head');
require('react');
require('@tanstack/react-query');
const indexBrowser = require('./chunks/index-browser.cjs');
require('next/router');
const _blitz = require('.blitz');
require('hoist-non-react-statics');
require('superjson');
require('blitz');
require('debug');



exports.BlitzProvider = indexBrowser.BlitzProvider;
exports.ErrorBoundary = indexBrowser.ErrorBoundary;
exports.ErrorComponent = indexBrowser.ErrorComponent;
exports.NoPageFlicker = indexBrowser.NoPageFlicker;
exports.RouterContext = indexBrowser.RouterContext;
exports.extractQueryFromAsPath = indexBrowser.extractQueryFromAsPath;
exports.extractRouterParams = indexBrowser.extractRouterParams;
exports.setupBlitzClient = indexBrowser.setupBlitzClient;
exports.useErrorHandler = indexBrowser.useErrorHandler;
exports.useParam = indexBrowser.useParam;
exports.useParams = indexBrowser.useParams;
exports.useRouterQuery = indexBrowser.useRouterQuery;
exports.withErrorBoundary = indexBrowser.withErrorBoundary;
exports.Routes = _blitz.Routes;
