// blitz/dist/chunks/dev.cjs
'use strict';

const arg = require('arg');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e["default"] : e; }

const arg__default = /*#__PURE__*/_interopDefaultLegacy(arg);

var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
const dev = () => __async(void 0, null, function* () {
  const nextArgs = arg__default(
    {
      "--port": Number,
      "--hostname": String,
      "--inspect": Boolean,
      "-p": "--port",
      "-H": "--hostname"
    },
    {
      permissive: true
    }
  );
  const config = {
    rootFolder: process.cwd(),
    port: nextArgs["--port"],
    hostname: nextArgs["--hostname"],
    inspect: nextArgs["--inspect"],
    env: process.env.NODE_ENV === "production" ? "prod" : "dev"
  };
  yield import('./next-commands.cjs').then((i) => i.dev(config));
});

exports.dev = dev;
