// blitz/dist/chunks/index2.cjs
'use strict';

const arg = require('arg');
const spawn = require('cross-spawn');
const fs = require('fs');
const path = require('path');
const dotenv = require('dotenv');
const dotenvExpand = require('dotenv-expand');
const fs$1 = require('fs-extra');
const pkgDir = require('pkg-dir');
const resolveCwd = require('resolve-cwd');
const resolveFrom = require('resolve-from');
const helpers = require('./helpers.cjs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e["default"] : e; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  const n = Object.create(null);
  if (e) {
    for (const k in e) {
      n[k] = e[k];
    }
  }
  n["default"] = e;
  return n;
}

const arg__default = /*#__PURE__*/_interopDefaultLegacy(arg);
const spawn__default = /*#__PURE__*/_interopDefaultLegacy(spawn);
const fs__namespace = /*#__PURE__*/_interopNamespace(fs);
const fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
const path__namespace = /*#__PURE__*/_interopNamespace(path);
const path__default = /*#__PURE__*/_interopDefaultLegacy(path);
const dotenv__namespace = /*#__PURE__*/_interopNamespace(dotenv);
const dotenvExpand__default = /*#__PURE__*/_interopDefaultLegacy(dotenvExpand);
const pkgDir__default = /*#__PURE__*/_interopDefaultLegacy(pkgDir);
const resolveCwd__default = /*#__PURE__*/_interopDefaultLegacy(resolveCwd);
const resolveFrom__default = /*#__PURE__*/_interopDefaultLegacy(resolveFrom);

let combinedEnv = void 0;
let cachedLoadedEnvFiles = [];
function processEnv(loadedEnvFiles, dir, log = console) {
  var _a;
  if (process.env.__NEXT_PROCESSED_ENV || loadedEnvFiles.length === 0) {
    return process.env;
  }
  process.env.__NEXT_PROCESSED_ENV = "true";
  const origEnv = Object.assign({}, process.env);
  const parsed = {};
  for (const envFile of loadedEnvFiles) {
    try {
      let result = {};
      result.parsed = dotenv__namespace.parse(envFile.contents);
      result = dotenvExpand__default.expand(result);
      if (result.parsed) {
        log.info(`Loaded env from ${path__namespace.join(dir || "", envFile.path)}`);
      }
      for (const key of Object.keys(result.parsed || {})) {
        if (typeof parsed[key] === "undefined" && typeof origEnv[key] === "undefined") {
          parsed[key] = (_a = result.parsed) == null ? void 0 : _a[key];
        }
      }
    } catch (err) {
      log.error(`Failed to load env from ${path__namespace.join(dir || "", envFile.path)}`, err);
    }
  }
  return Object.assign(process.env, parsed);
}
function loadEnvConfig(dir = process.cwd(), _dev, log = console, { ignoreCache } = { ignoreCache: false }) {
  var _a, _b;
  if (combinedEnv && !ignoreCache)
    return { combinedEnv, loadedEnvFiles: cachedLoadedEnvFiles };
  const appEnv = (_b = (_a = process.env.APP_ENV) != null ? _a : process.env.NODE_ENV) != null ? _b : "development";
  let dotenvFiles = [
    `.env.${appEnv}.local`,
    `.env.${appEnv}`,
    appEnv !== "test" && `.env.local`,
    ".env"
  ].filter(Boolean);
  for (const envFile of dotenvFiles) {
    const dotEnvPath = path__namespace.join(dir, envFile);
    try {
      const stats = fs__namespace.statSync(dotEnvPath);
      if (!stats.isFile()) {
        continue;
      }
      const contents = fs__namespace.readFileSync(dotEnvPath, "utf8");
      cachedLoadedEnvFiles.push({
        path: envFile,
        contents
      });
    } catch (err) {
      if (err.code !== "ENOENT") {
        log.error(`Failed to load env from ${envFile}`, err);
      }
    }
  }
  combinedEnv = processEnv(cachedLoadedEnvFiles, dir, log);
  return { combinedEnv, loadedEnvFiles: cachedLoadedEnvFiles };
}

const NON_STANDARD_NODE_ENV = `You are using a non-standard "NODE_ENV" value in your environment. This creates inconsistencies in the project and is strongly advised against. Read more: https://nextjs.org/docs/messages/non-standard-node-env`;

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __async$1 = (__this, __arguments, generator) => {
  return new Promise((resolve2, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve2(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
const debug = require("debug")("blitz:utils");
function resolveBinAsync(_0) {
  return __async$1(this, arguments, function* (pkg, executable = pkg) {
    const packageDir = yield pkgDir__default(resolveCwd__default(pkg));
    if (!packageDir)
      throw new Error(`Could not find package.json for '${pkg}'`);
    const { bin } = yield fs$1.readJSON(path__default.join(packageDir, "package.json"));
    const binPath = typeof bin === "object" ? bin[executable] : bin;
    if (!binPath)
      throw new Error(`No bin '${executable}' in module '${pkg}'`);
    debug("binPath: " + binPath);
    const fullPath = path__default.join(packageDir, binPath);
    debug("fullPath: " + fullPath);
    return fullPath;
  });
}
function isInDocker() {
  const cgroupFile = path.join("proc", "self", "cgroup");
  if (fs__default.existsSync(cgroupFile)) {
    const content = fs__default.readFileSync(cgroupFile, "utf-8");
    return content.includes("docker");
  }
  return false;
}
const defaults = {
  hostname: isInDocker() ? "0.0.0.0" : "127.0.0.1",
  env: "prod"
};
function normalize(config) {
  return __async$1(this, null, function* () {
    var _a, _b, _c;
    const rootFolder = path.resolve(process.cwd(), config.rootFolder);
    const env = config.env || defaults.env;
    return __spreadProps(__spreadValues({}, config), {
      hostname: (_a = config.hostname) != null ? _a : defaults.hostname,
      env,
      rootFolder,
      isTypeScript: (_b = config.isTypeScript) != null ? _b : yield getIsTypeScript(rootFolder),
      watch: (_c = config.watch) != null ? _c : env === "dev",
      clean: config.clean,
      nextBin: yield getCommandBin("next", rootFolder, env === "dev")
    });
  });
}
function getCommandBin(_0) {
  return __async$1(this, arguments, function* (command, rootFolder = process.cwd(), _usePatched = false) {
    const bin = yield resolveBinAsync(command);
    return path.resolve(rootFolder, bin);
  });
}
function getIsTypeScript(rootFolder) {
  return __async$1(this, null, function* () {
    try {
      yield fs.promises.access(path.join(rootFolder, "tsconfig.json"));
      return true;
    } catch (e) {
      return false;
    }
  });
}

const globalBlitzPath = resolveFrom__default.silent(__dirname, "blitz");
const localBlitzPath = resolveFrom__default.silent(process.cwd(), "blitz");
const localBlitzAuthPath = resolveFrom__default.silent(process.cwd(), "@blitzjs/auth");
const localBlitzRpcPath = resolveFrom__default.silent(process.cwd(), "@blitzjs/rpc");
const localBlitzNextPath = resolveFrom__default.silent(process.cwd(), "@blitzjs/next");
function readVersions() {
  const globalBlitzPkgJsonPath = pkgDir__default.sync(globalBlitzPath);
  const localBlitzPkgJsonPath = pkgDir__default.sync(localBlitzPath);
  const localBlitzAuthPkgJsonPath = pkgDir__default.sync(localBlitzAuthPath);
  const localBlitzNextPkgJsonPath = pkgDir__default.sync(localBlitzNextPath);
  const localBlitzRpcPkgJsonPath = pkgDir__default.sync(localBlitzRpcPath);
  const versions = { globalVersion: "", localVersions: {} };
  if (globalBlitzPkgJsonPath && globalBlitzPkgJsonPath !== localBlitzPkgJsonPath) {
    versions.globalVersion = require(path.join(globalBlitzPkgJsonPath, "package.json")).version;
  }
  if (localBlitzPkgJsonPath) {
    versions.localVersions.blitz = require(path.join(localBlitzPkgJsonPath, "package.json")).version;
  }
  if (localBlitzAuthPkgJsonPath) {
    versions.localVersions.blitzAuth = require(path.join(
      localBlitzAuthPkgJsonPath,
      "package.json"
    )).version;
  }
  if (localBlitzNextPkgJsonPath) {
    versions.localVersions.blitzNext = require(path.join(
      localBlitzNextPkgJsonPath,
      "package.json"
    )).version;
  }
  if (localBlitzRpcPkgJsonPath) {
    versions.localVersions.blitzRpc = require(path.join(
      localBlitzRpcPkgJsonPath,
      "package.json"
    )).version;
  }
  return versions;
}

var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
const commonArgs = {
  "--version": Boolean,
  "--help": Boolean,
  "--inspect": Boolean,
  "--env": String,
  "-v": "--version",
  "-h": "--help",
  "-e": "--env"
};
const args = arg__default(commonArgs, {
  permissive: true
});
const commands = {
  dev: () => import('./dev.cjs').then((i) => i.dev),
  build: () => import('./build.cjs').then((i) => i.build),
  start: () => import('./start.cjs').then((i) => i.start),
  export: () => import('./export.cjs').then((i) => i.nextExport),
  new: () => import('./new.cjs').then((i) => i.newApp),
  generate: () => import('./generate.cjs').then((i) => i.generate),
  codegen: () => import('./codegen.cjs').then((i) => i.codegen),
  db: () => import('./db.cjs').then((i) => i.db),
  install: () => import('./install.cjs').then((i) => i.install),
  console: () => import('./console.cjs').then((i) => i.consoleREPL),
  routes: () => import('./routes.cjs').then((i) => i.routes)
};
const aliases = {
  d: "dev",
  b: "build",
  s: "start",
  e: "export",
  n: "new",
  g: "generate",
  i: "install",
  c: "console",
  r: "routes"
};
let blitzCommand;
if (commands[args._[0]]) {
  blitzCommand = args._[0];
}
if (aliases[args._[0]]) {
  blitzCommand = aliases[args._[0]];
}
const forwardedArgs = blitzCommand ? args._.slice(1) : args._;
function runCommandFromBin() {
  return __async(this, null, function* () {
    if (!args._[0]) {
      console.log("No command specified");
      process.exit(1);
    }
    let commandBin = null;
    try {
      commandBin = yield getCommandBin(args._[0]);
    } catch (e) {
      console.error(`Error: ${e.message}`);
    }
    if (!commandBin) {
      process.exit(1);
    }
    const result = spawn__default.sync(commandBin, args._.slice(1), { stdio: "inherit" });
    process.exit(result.status || 0);
  });
}
function printEnvInfo() {
  return __async(this, null, function* () {
    const osName = yield import('os-name');
    const envinfo = yield import('envinfo');
    const pkgManager = helpers.getPkgManager();
    const env = yield envinfo.default.run(
      {
        System: ["OS", "CPU", "Memory", "Shell"],
        Binaries: ["Node", "Yarn", "npm", "pnpm"],
        npmPackages: [
          "blitz",
          "@blitzjs/rpc",
          "@blitzjs/auth",
          "@blitzjs/next",
          "typescript",
          "react",
          "react-dom",
          "prisma",
          "@prisma/client",
          "next"
        ]
      },
      { showNotFound: true }
    );
    const { globalVersion, localVersions } = readVersions();
    if (globalVersion) {
      console.log(`Blitz version: ${globalVersion} (global)`);
    }
    if (localVersions.blitz) {
      console.log(`Blitz version: ${localVersions.blitz} (local)`);
    }
    console.log(
      `${osName.default()} | ${process.platform}-${process.arch} | Node: ${process.version}
`
    );
    console.log(`
 Package manager: ${pkgManager}`);
    console.log(env);
    process.exit(0);
  });
}
function main() {
  return __async(this, null, function* () {
    if (args["--env"]) {
      process.env.APP_ENV = args["--env"];
    }
    if (args["_"].length === 0 && args["--version"]) {
      yield printEnvInfo();
    }
    if (args["--help"]) {
      forwardedArgs.push("--help");
    }
    const defaultEnv = blitzCommand === "build" || blitzCommand === "start" ? "production" : "development";
    const standardEnv = ["production", "development", "test"];
    if (process.env.NODE_ENV && !standardEnv.includes(process.env.NODE_ENV)) {
      console.warn(NON_STANDARD_NODE_ENV);
    }
    process.env.NODE_ENV = process.env.NODE_ENV || defaultEnv;
    loadEnvConfig(process.cwd(), void 0, { error: console.error, info: console.info });
    process.on("SIGTERM", () => process.exit(0));
    process.on("SIGINT", () => process.exit(0));
    if (blitzCommand) {
      const commandFn = commands[blitzCommand];
      commandFn == null ? void 0 : commandFn().then((exec) => exec(forwardedArgs)).then(() => {
        if (blitzCommand === "build") {
          process.exit(0);
        }
      }).catch((err) => {
        console.log(err);
      });
    } else {
      if (args["--help"] && forwardedArgs.length === 1 && forwardedArgs[0] === "--help") {
        console.log(`
      Usage
        $ blitz <command>
  
      Available commands
        dev, d          Start a development server \u{1F984} 
        build, b        Create a production build \u{1F3D7}\uFE0F
        start, s        Start the production server \u{1F40E}
        export, e       Export application to static HTML \u{1F4C1}
        new, n          Create a new Blitz project \u2728
        generate, g     Generate new files for your Blitz project \u{1F920}
        codegen         Run the blitz codegen \u{1F916}
        db              Run database commands \u{1F5C4}\uFE0F
        
      Options
        --env, -e       App environment name
        --version, -v   Version number
        --help, -h      Displays this message
  
      For more information run a command with the --help flag
        $ blitz build --help
    `);
        process.exit(0);
      } else {
        void runCommandFromBin();
      }
    }
  });
}
main().catch((e) => {
  console.error(e);
});

exports.loadEnvConfig = loadEnvConfig;
exports.normalize = normalize;
exports.readVersions = readVersions;
