// blitz/dist/chunks/logging.cjs
'use strict';

const tslog = require('tslog');
const chalk$1 = require('chalk');
const consoleTablePrinter = require('console-table-printer');
const ora = require('ora');
const readline = require('readline');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e["default"] : e; }

const chalk__default = /*#__PURE__*/_interopDefaultLegacy(chalk$1);
const ora__default = /*#__PURE__*/_interopDefaultLegacy(ora);
const readline__default = /*#__PURE__*/_interopDefaultLegacy(readline);

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
const baseLogger = (options = {}) => {
  if (globalThis._blitz_baseLogger)
    return globalThis._blitz_baseLogger;
  globalThis._blitz_baseLogger = BlitzLogger(options);
  return globalThis._blitz_baseLogger;
};
const BlitzLogger = (settings = {}) => {
  const baseLogger2 = new tslog.Logger(__spreadValues({
    minLevel: "info",
    type: "pretty",
    dateTimePattern: process.env.NODE_ENV === "production" ? "year-month-day hour:minute:second.millisecond" : "hour:minute:second.millisecond",
    displayFunctionName: false,
    displayFilePath: "hidden",
    displayRequestId: false,
    dateTimeTimezone: process.env.NODE_ENV === "production" ? "utc" : Intl.DateTimeFormat().resolvedOptions().timeZone,
    prettyInspectHighlightStyles: {
      name: "yellow",
      number: "blue",
      bigint: "blue",
      boolean: "blue"
    },
    maskValuesOfKeys: ["password", "passwordConfirmation"],
    exposeErrorCodeFrame: process.env.NODE_ENV !== "production"
  }, settings));
  return baseLogger2;
};
const initializeLogger = (logger) => {
  globalThis._blitz_baseLogger = logger;
};
const table = consoleTablePrinter.Table;
const chalk = chalk__default;
const blitzBrightBrandColor = "8a3df0";
const brandColor = blitzBrightBrandColor;
const newLine = () => {
  const logLevel = globalThis._blitz_logLevel;
  switch (logLevel) {
    case "trace":
    case "debug":
    case "info":
      console.log(" ");
      break;
  }
};
const withBrand = (str) => {
  return chalk__default.hex(brandColor).bold(str);
};
const withCaret = (str) => {
  return `${chalk__default.gray(">")} ${str}`;
};
const withCheck = (str) => {
  return `${chalk__default.green("\u2714")} ${str}`;
};
const withProgress = (str) => {
  return withCaret(str);
};
const greenText = (str) => {
  return `${chalk__default.green(str)}`;
};
const branded = (msg) => {
  console.log(chalk__default.hex(brandColor).bold(msg));
};
const clearLine = (msg) => {
  readline__default.clearLine(process.stdout, 0);
  readline__default.cursorTo(process.stdout, 0);
  msg && process.stdout.write(msg);
};
const clearConsole = () => {
  if (process.platform === "win32") {
    process.stdout.write("\x1B[2J\x1B[0f");
  } else {
    process.stdout.write("\x1B[2J\x1B[3J\x1B[H");
  }
};
const progress = (msg) => {
  console.log(withProgress(msg));
};
const spinner = (str) => {
  return ora__default({
    text: str,
    color: "blue",
    spinner: {
      interval: 120,
      frames: ["\u25E2", "\u25E3", "\u25E4", "\u25E5"]
    }
  });
};
const success = (msg) => {
  console.log(withCheck(chalk__default.green(msg)));
};
const error = (msg) => {
  console.log(`${chalk__default.red(msg)}`);
};
const variable = (val) => {
  return chalk__default.cyan.bold(`${val}`);
};
const debug = require("debug")("blitz");
const log = {
  withBrand,
  withCaret,
  branded,
  clearLine,
  clearConsole,
  progress,
  spinner,
  success,
  greenText,
  error,
  variable,
  debug,
  Table: consoleTablePrinter.Table
};

exports.BlitzLogger = BlitzLogger;
exports.baseLogger = baseLogger;
exports.chalk = chalk;
exports.initializeLogger = initializeLogger;
exports.log = log;
exports.newLine = newLine;
exports.table = table;
