// blitz/dist/chunks/logging.mjs
import { Logger } from 'tslog';
import chalk$1 from 'chalk';
import { Table } from 'console-table-printer';
import ora from 'ora';
import readline from 'readline';

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
const baseLogger = (options = {}) => {
  if (globalThis._blitz_baseLogger)
    return globalThis._blitz_baseLogger;
  globalThis._blitz_baseLogger = BlitzLogger(options);
  return globalThis._blitz_baseLogger;
};
const BlitzLogger = (settings = {}) => {
  const baseLogger2 = new Logger(__spreadValues({
    minLevel: "info",
    type: "pretty",
    dateTimePattern: process.env.NODE_ENV === "production" ? "year-month-day hour:minute:second.millisecond" : "hour:minute:second.millisecond",
    displayFunctionName: false,
    displayFilePath: "hidden",
    displayRequestId: false,
    dateTimeTimezone: process.env.NODE_ENV === "production" ? "utc" : Intl.DateTimeFormat().resolvedOptions().timeZone,
    prettyInspectHighlightStyles: {
      name: "yellow",
      number: "blue",
      bigint: "blue",
      boolean: "blue"
    },
    maskValuesOfKeys: ["password", "passwordConfirmation"],
    exposeErrorCodeFrame: process.env.NODE_ENV !== "production"
  }, settings));
  return baseLogger2;
};
const initializeLogger = (logger) => {
  globalThis._blitz_baseLogger = logger;
};
const table = Table;
const chalk = chalk$1;
const blitzBrightBrandColor = "8a3df0";
const brandColor = blitzBrightBrandColor;
const newLine = () => {
  const logLevel = globalThis._blitz_logLevel;
  switch (logLevel) {
    case "trace":
    case "debug":
    case "info":
      console.log(" ");
      break;
  }
};
const withBrand = (str) => {
  return chalk$1.hex(brandColor).bold(str);
};
const withCaret = (str) => {
  return `${chalk$1.gray(">")} ${str}`;
};
const withCheck = (str) => {
  return `${chalk$1.green("\u2714")} ${str}`;
};
const withProgress = (str) => {
  return withCaret(str);
};
const greenText = (str) => {
  return `${chalk$1.green(str)}`;
};
const branded = (msg) => {
  console.log(chalk$1.hex(brandColor).bold(msg));
};
const clearLine = (msg) => {
  readline.clearLine(process.stdout, 0);
  readline.cursorTo(process.stdout, 0);
  msg && process.stdout.write(msg);
};
const clearConsole = () => {
  if (process.platform === "win32") {
    process.stdout.write("\x1B[2J\x1B[0f");
  } else {
    process.stdout.write("\x1B[2J\x1B[3J\x1B[H");
  }
};
const progress = (msg) => {
  console.log(withProgress(msg));
};
const spinner = (str) => {
  return ora({
    text: str,
    color: "blue",
    spinner: {
      interval: 120,
      frames: ["\u25E2", "\u25E3", "\u25E4", "\u25E5"]
    }
  });
};
const success = (msg) => {
  console.log(withCheck(chalk$1.green(msg)));
};
const error = (msg) => {
  console.log(`${chalk$1.red(msg)}`);
};
const variable = (val) => {
  return chalk$1.cyan.bold(`${val}`);
};
const debug = require("debug")("blitz");
const log = {
  withBrand,
  withCaret,
  branded,
  clearLine,
  clearConsole,
  progress,
  spinner,
  success,
  greenText,
  error,
  variable,
  debug,
  Table
};

export { BlitzLogger as B, baseLogger as b, chalk as c, initializeLogger as i, log as l, newLine as n, table as t };
