// blitz/dist/chunks/start.mjs
import arg from 'arg';

var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
const start = () => __async(void 0, null, function* () {
  const nextArgs = arg(
    {
      "--port": Number,
      "--hostname": String,
      "--inspect": Boolean,
      "-p": "--port",
      "-H": "--hostname"
    },
    {
      permissive: true
    }
  );
  const config = {
    rootFolder: process.cwd(),
    port: nextArgs["--port"],
    hostname: nextArgs["--hostname"],
    inspect: nextArgs["--inspect"],
    env: process.env.NODE_ENV === "production" ? "prod" : "dev"
  };
  yield import('./next-commands.mjs').then((i) => i.prod(config));
});

export { start };
