// @blitzjs/rpc/dist/loader-server-resolvers.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const loaderUtils = require('./chunks/loader-utils.cjs');
const rpc = require('./chunks/rpc.cjs');
require('@tanstack/react-query');
require('blitz');
require('next/router');
const path = require('path');
require('next/dist/client/normalize-trailing-slash');
require('next/dist/client/add-base-path');
require('superjson');
require('@blitzjs/auth');

var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
function loader(input) {
  return __async(this, null, function* () {
    const compiler = this._compiler;
    const id = this.resource;
    const root = this._compiler.context;
    const isSSR = compiler.name === "server";
    if (isSSR) {
      return yield transformBlitzRpcResolverServer(
        input,
        loaderUtils.toPosixPath(id),
        loaderUtils.toPosixPath(root),
        this.query
      );
    }
    return input;
  });
}
module.exports = loader;
function transformBlitzRpcResolverServer(src, id, root, options) {
  return __async(this, null, function* () {
    loaderUtils.assertPosixPath(id);
    loaderUtils.assertPosixPath(root);
    const resolverFilePath = "/" + path.posix.relative(root, id);
    loaderUtils.assertPosixPath(resolverFilePath);
    const routePath = loaderUtils.convertPageFilePathToRoutePath({
      appRoot: root,
      absoluteFilePath: resolverFilePath,
      resolverBasePath: options == null ? void 0 : options.resolverPath,
      extraRpcBasePaths: options == null ? void 0 : options.includeRPCFolders
    });
    const resolverName = loaderUtils.convertFilePathToResolverName(resolverFilePath);
    const resolverType = loaderUtils.convertFilePathToResolverType(resolverFilePath);
    const fullRoutePath = rpc.normalizeApiRoute("/api/rpc" + routePath);
    const lines = src.split("\n");
    const newLines = lines.map((line) => {
      if (line.trim().startsWith("export default")) {
        return line.replace("export default", "const __internal_rpcHandler =");
      }
      return line;
    });
    return `${newLines.join("\n")}

__internal_rpcHandler._resolverName = '${resolverName}'
__internal_rpcHandler._resolverType = '${resolverType}'
__internal_rpcHandler._routePath = '${fullRoutePath}'

export default __internal_rpcHandler`;
  });
}

exports.loader = loader;
exports.transformBlitzRpcResolverServer = transformBlitzRpcResolverServer;
